text/0161-telemetry-schema-and-resource.md

```diff
@@ -0,0 +1,196 @@
+TelemetrySchema version for a stable API.
+
+## Motivation
+
+In the [current definition](https://github.com/open-telemetry/opentelemetry-specification/pull/1692)
+of Resource + Telemetry schema interaction, the specification focuses mostly on
+how to provide a consistent "storage" of Resource within the SDK. However,
+from a usage standpoint, how do consumers gain stability across changes to
+schema?
+
+- Telemetry Schema allows the notion of "renaming" attributes in a schema.
+  Can a consumer of Resource "lock" in a specific version of the schema to
+  ensure no renaming occurs?
+- If resource detection is provided by libraries external to open-telemetry, do
+  we require ALL resource detectors to be against the same telemetry schema
+  version number?  This would inhibit upgrading versions as you'd need to wait
+  for all resource detectors to upgrade, even if a (stable) conversion exists
+  from the resource version.
+
+Specifically, we need a way for exporters that rely on Resource labels to ensure
+various version differences across OpenTelemetry components do not cause
+breakages by leveraging the powers provided by telemetry schema.
+
+As an example, prior to schema version and stability, the PR
+[#1495](https://github.com/open-telemetry/opentelemetry-specification/pull/1495)
+renamed `cloud.zone` to `cloud.availability_zone`.  In a Schema world, this
+change could have been defined with the following schema rule:
+
+```yaml
+  next.version:
+    resources:
+      changes:
+        - rename_attributes:
+            attribute_map:
+              cloud.zone: cloud.availability_zone
+  previous.version:
+```
+
+If an exporter is defined expecting to find `cloud.zone` on `previous_version`,
+then when the SDK is bumped to `next.version` it will no longer be able to see
+`cloud.zone`.
+
+Specifically, we assume the following components could have independent
+versioning + release cycles:
+
+```
++-------------------+     +-----+     +----------+
+| Resource Detector | <-> | SDK | <-> | Exporter |
++-------------------+     +-----+     +----------+
+```
+
+## Explanation
+
+We propose a new set of SDK features:
+
+- A mechanism to "migrate" a resource up/down compatible schema changes.
+- A mechanism to determine if two schema URLs allow compatible changes.
+
+### Consuming resources at specific schema version
+
+The first is done by providing a new method against `Resource` in the SDK that
+will transform the attributes as necessary to match a given schema version.
+
+e.g. In Java you might have the following for resource:
+
+```java
+public abstract class Resource {
+    ... existing methods ...
+    /** The current schema version. */
+    public String getSchemaUrl();
+    /**
+     * Returns a new instance of Resource with labels compatible with the
+     * given schema url.
+     *
+     * <p> Returns an invalid Resource if conversion is not possible.
+     */
+    public Resource convertToSchema(String schemaUrl);
```

Comment:
Nice! This is an expected capability of schemas, so no surprises here.
I believe this should also be able to return an error since some conversions are not possible (e.g. between different schema families or if the versions are incompatible).

Reply:
Since Resource resolution happens at application startup, I'm nervous about having SDKs have to go and fetch multiple schemas from the public internet as a part of the startup process, in order to perform the conversion. There are failure modes here, and general startup-time issues to be considered (especially, for example, in mobile or FAAS implementations).

Reply:
> Since Resource resolution happens at application startup, I'm nervous about having SDKs have to go and fetch multiple schemas from the public internet as a part of the startup process, in order to perform the conversion.

That's a good point. The SDKs can have the schema file that they correspond to embedded with the source code so that there is no need to fetch anything for detectors that are part of the SDK. If each SDK includes the latest Otel schema file then any Otel schema conversions can happen without network access.
Custom, non-Otel schema conversions will still require a network fetching. Perhaps we also provide an alternate `convertToSchema` function which accepts a schema file instead of schema URL?

Reply:
I don't think that an alternate conversion method should be required to work from a file, but an alternate URL->bytes function could be provided.  I'd imagine possibly a three tier lookup: 1) embedded in the application by the SDK, 2) on disk in a well-known or specified location 3) over the network through traditional URL resolution.

Reply:
Per this thread conversation, it may be better to start with a solution that "file"/"bytes" for the schema are provided, and delegate to the user of the API to ensure that is available somewhere to grab. Some users may embed the files in their application, some may embed the schema bytes, etc.

@jkwatson had a very good point, and probably we should postpone the API that fetches from the internet.

Reply:
I agree with the notion of providing the local schema instead of fetching from the URL. We may go a step further and define `Schema` as an opaque object that is loadable from file or from a URL. This opens up the possibility of performance improvements (parse and compile the schema once such that the subsequent conversion calls are fast). See e.g. [prototype here](https://github.com/tigrannajaryan/telemetry-schema/tree/main/schema) that has "parse/compile" and "convert" as separate operations.